src/operations/client_bulk_write/command_builder.ts

```diff
@@ -0,0 +1,233 @@
+   * user provided verboseResults option. Defaults to true.
+   */
+  get errorsOnly(): boolean {
+    if ('verboseResults' in this.options) {
+      return !this.options.verboseResults;
+    }
+    return true;
+  }
+
+  /**
+   * Build the bulk write commands from the models.
+   */
+  buildCommands(): ClientBulkWriteCommand[] {
+    // Iterate the models to build the ops and nsInfo fields.
+    const operations = [];
+    let currentNamespaceIndex = 0;
+    const namespaces = new Map<string, number>();
+    for (const model of this.models) {
+      const ns = model.namespace;
+      if (namespaces.has(ns)) {
+        operations.push(buildOperation(model, namespaces.get(ns) as number));
+      } else {
+        namespaces.set(ns, currentNamespaceIndex);
+        operations.push(buildOperation(model, currentNamespaceIndex));
+        currentNamespaceIndex++;
+      }
+    }
+
+    const nsInfo = Array.from(namespaces.keys()).map(ns => {
+      return { ns: ns };
+    });
+
+    // The base command.
+    const command: ClientBulkWriteCommand = {
+      bulkWrite: 1,
+      errorsOnly: this.errorsOnly,
+      ordered: this.options.ordered ?? true,
+      ops: new DocumentSequence(operations),
+      nsInfo: new DocumentSequence(nsInfo)
+    };
+    // Add bypassDocumentValidation if it was present in the options.
+    if ('bypassDocumentValidation' in this.options) {
+      command.bypassDocumentValidation = this.options.bypassDocumentValidation;
+    }
+    // Add let if it was present in the options.
+    if ('let' in this.options) {
+      command.let = this.options.let;
+    }
+    return [command];
+  }
+}
+
+/**
+ * Build the insert one operation.
+ * @param model - The insert one model.
+ * @param index - The namespace index.
+ * @returns the operation.
+ */
+export const buildInsertOneOperation = (model: ClientInsertOneModel, index: number): Document => {
+  const document: Document = {
+    insert: index,
+    document: model.document
+  };
+  return document;
+};
+
+/**
+ * Build the delete one operation.
+ * @param model - The insert many model.
+ * @param index - The namespace index.
+ * @returns the operation.
+ */
+export const buildDeleteOneOperation = (model: ClientDeleteOneModel, index: number): Document => {
+  return createDeleteOperation(model, index, false);
+};
+
+/**
+ * Build the delete many operation.
+ * @param model - The delete many model.
+ * @param index - The namespace index.
+ * @returns the operation.
+ */
+export const buildDeleteManyOperation = (model: ClientDeleteManyModel, index: number): Document => {
+  return createDeleteOperation(model, index, true);
+};
+
+/**
+ * Creates a delete operation based on the parameters.
+ */
+function createDeleteOperation(
+  model: ClientDeleteOneModel | ClientDeleteManyModel,
+  index: number,
+  multi: boolean
+): Document {
+  const document: Document = {
+    delete: index,
+    multi: multi,
+    filter: model.filter
+  };
```

Comment:
```suggestion
  const document: {
    delete: number;
    multi: boolean;
    filter: Filter<TSchema>;
    hint?: Hint;
    collation?: CollationOptions;
  } = {
    delete: index,
    multi: multi,
    filter: model.filter
  };
```

Typing all the commands accurately doesn't seem to introduce any hurdles to overcome and would help future maintainers to make sure the optional assignments (like `if(hint)`) remain correct.

Reply:
I've added types now for all the operations.